Blackbox test improvements for Jenkins

dgonyeo opened the pull request:
This contains a set of changes/improvements to the blackbox tests I wrote in service of getting Jenkins to work with Ignition.

dgonyeo commented:
This is ready for review now. I've updated the title/body of the PR to reflect what this ended up becoming.

arithx commented on tests/types/types.go:
```diff
@@ -172,31 +172,6 @@ func GetBaseDisk() []Disk {
 					Length:         262144,
 					FilesystemType: "vfat",
 					Hybrid:         true,
-					Files: []File{
```
If we're going to remove these from the baseline checking can we add these files as validation in the `In` & `Out` partitions of the `EmptyUserdata` test (or a similar test) to make sure that we don't introduce a regression where files are removed that shouldn't be?

arithx commented on tests/blackbox_test.go:
```diff
@@ -170,8 +176,16 @@ func outer(t *testing.T, test types.Test, negativeTests bool) {
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(oemLookasideDir)
```
Having the defer after the error block can potentially leave files/directories existing.

arithx commented on tests/blackbox_test.go:
```diff
@@ -184,28 +198,87 @@ func outer(t *testing.T, test types.Test, negativeTests bool) {
+		// Move value into the local scope, because disk.ImageFile will change
+		// by the time this runs
+		imageFile := disk.ImageFile
+		defer func() {
```
Same as earlier, we could create the file & truncate it to the full size but fail the partition table creation and then error out and leave this not cleaned up.

arithx commented on tests/blackbox_test.go:
```diff
@@ -184,28 +198,87 @@ func outer(t *testing.T, test types.Test, negativeTests bool) {
+			return err
+		}
+		device := disk.Device
+		defer func() {
```
If there's an error partitioning the device then this after the error block will not run `losetup -d`

arithx commented on tests/blackbox_test.go:
```diff
@@ -184,28 +198,87 @@ func outer(t *testing.T, test types.Test, negativeTests bool) {
+			return err
+		}
+		partitions := disk.Partitions
+		defer func() {
```
If `systemd.WaitOnDevice` fails in `mountRootPartition` then the mount point will be leaked

ajeddeloh commented on tests/filesystem.go:
```diff
@@ -250,18 +251,31 @@ func mountRootPartition(t *testing.T, partitions []*types.Partition) bool {
 		_ = mustRun(t, "mount", args...)
+
+		t.Logf("starting waiting on %s", partition.Device)
+		err := systemd.WaitOnDevices([]string{partition.Device}, "tests")
```
You're waiting on the device your mounting after you mount it?

dgonyeo replied:
Is this unnecessary? The btrfs tests pass without this check.

ajeddeloh replied:
This makes them fail?

dgonyeo replied:
This makes them fail any runs after the first one: https://gist.github.com/dgonyeo/21de88e8eda1ae68cef8f935b294a0c5

dgonyeo replied:
If you've got ideas as to why I'd be happy to investigate them

ajeddeloh commented on tests/blackbox_test.go:
```diff
@@ -375,7 +364,23 @@ func outer(t *testing.T, test types.Test, negativeTests bool) error {
 			}
-			validateFilesDirectoriesAndLinks(t, disk.Partitions)
+
+			if needsNodeValidation(disk.Partitions) {
```
We should fix the problem with the btrfs tests not work around them.

dgonyeo commented:
Update for anyone watching this: this is blocked until we figure out what's going on with the btrfs test. Issue to track the problem exposed is here: https://github.com/coreos/bugs/issues/2328

dgonyeo commented:
After talking with @crawford he doesn't believe that bug is worth our time investigating, as systemd device units were never intended as something that should be manually started. I've removed the unnecessary device units from these tests, and this PR should be ready for another round of reviews.

ajeddeloh commented:
Can you remove the part where we don't validate everything? IIRC if you got rid of the device units it wasn't neceessary, right?

dgonyeo commented:
That was an optimization. I can take it out if you want, but it only doesn't validate files if there are no files to validate. There are still tests with files in them, and file validation occurs in them. I also added a test for checking that the specific files in the efi partition that I took out of the base disk remain unmodified by Ignition.

dgonyeo commented:
@arithx done

ajeddeloh commented:
I'm still inclined to remove that optimization. Even if there are not any files, dirs, etc, it does ensure the partitions are mountable. There's also the files in the base disk which we test aren't being clobbered. Furthermore if we add something else we may want to validate we don't have to worry about it being forgotten in `needsNodeValidation()`.

dgonyeo commented:
Just removed the commit with that change from this PR.

dgonyeo commented:
Oh I also just removed the test I added for the files on the efi partition, since they're back in the base disk they'll be checked for with every test, making that particular test redundant.
